crates/task-impls/src/quorum_vote.rs

```diff
@@ -70,12 +74,170 @@ struct VoteDependencyHandle<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+                self.version,
+            )
+            .await
+            .context("Block header doesn't extend the proposal!")?;
+
+        let state = Arc::new(validated_state);
+        let delta = Arc::new(state_delta);
+
+        // TODO Upgrades
+        // Validate the DAC.
+        // let message = if cert.is_valid_cert(Upgrade Certificate) {
+        //     // Validate the block payload commitment for non-genesis DAC.
+        //     if cert.date().payload_commit != proposal.block_header.payload_commitment() {
+        //         warn!(
+        //             "Block payload commitment does not equal da cert payload commitment. View = {}",
+        //             *view
+        //         );
+        //         return false;
+        //     }
+        //     if let Ok(vote) = QuorumVote::<TYPES>::create_signed_vote(
+        //         QuorumData {
+        //             leaf_commit: proposed_leaf.commit(),
+        //         },
+        //         view,
+        //         &public_key,
+        //         &self.private_key
+        //     ) {
+        //         GeneralConsensusMessage::<TYPES>::Vote(vote)
+        //     } else {
+        //         error!("Unable to sign quorum vote!");
+        //         return false;
+        //     }
+        // } else {
+        //     error!(
+        //         "Invalid DAC in proposal! Skipping proposal. {:?} cur view is: {:?}",
+        //         cert, cur_view
+        //     );
+        //     return false;
+        // };
+
+        // Now that we've rounded everyone up, we need to update the shared state and broadcast our events.
+        // We will defer broadcast until all states are updated to avoid holding onto the lock during a network call.
+        let mut consensus_writer = self.consensus.write().await;
+
+        // Note: There's a potential inconsistency here between `self.view_number` and `leaf.data.view_number()`.
+        let view = View {
+            view_inner: ViewInner::Leaf {
+                leaf: proposed_leaf.commit(),
+                state: Arc::clone(&state),
+                delta: Some(Arc::clone(&delta)),
+            },
+        };
+        consensus_writer.update_validated_state_map(self.view_number, view.clone());
+        consensus_writer.update_saved_leaves(proposed_leaf.clone());
+
+        // Kick back our updated structures for downstream usage.
+        let new_leaves = consensus_writer.saved_leaves().clone();
+        let new_state = consensus_writer.validated_state_map().clone();
+        drop(consensus_writer);
+
+        // Broadcast now that the lock is dropped.
+        broadcast_event(
+            HotShotEvent::ValidatedStateUpdated(self.view_number, view).into(),
+            &self.sender,
+        )
+        .await;
+
+        // Send the new state up to the sequencer.
+        self.storage
+            .write()
+            .await
+            .update_undecided_state(new_leaves, new_state)
+            .await?;
```

Comment:
I think we might want to do this before broadcasting the event so we only act on shared state updates if it's persisted, but I don't actually think it's that important

Reply:
We get into a chicken and egg situation somewhat since this event is fallible. So our state update goes fine, then this could kill our code path while leaving the state updated on the backend.

Reply:
I think that situation is ok.  Basically it would mean we won't vote, but we would still have the proposal in our storage if we restart so we could vote on a later proposal.  It's not at all critical though so we can leave as is